qiskit/circuit/equivalence.py

```diff
@@ -62,13 +120,27 @@ def add_equivalence(self, gate, equivalent_circuit):
         _raise_if_param_mismatch(gate.params, equivalent_circuit.parameters)
 
         key = Key(name=gate.name, num_qubits=gate.num_qubits)
-
         equiv = Equivalence(params=gate.params.copy(), circuit=equivalent_circuit.copy())
 
-        if key not in self._map:
-            self._map[key] = Entry(search_base=True, equivalences=[])
+        target = self._set_default_node(key)
+        self._graph[target]["equivs"].append(equiv)
+
+        sources = {
+            Key(name=instruction.operation.name, num_qubits=len(instruction.qubits))
+            for instruction in equivalent_circuit
+        }
+        edges = [
+            (
+                self._set_default_node(source),
+                target,
+                {"index": self._rule_count, "rule": equiv, "source": source},
```

Comment:
Like the removed `Entry`, this is another place where it would be nicer to use proper structured data rather than defining an ad-hoc dictionary schema here in the middle of a function.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/8790/commits/73b553615a6b84098b67a6f53daa7c84f4cce872.